### DISCUSSION TITLE
MET 11.0.0 install error: fatal error: netcdf: No such file or directory

### DISCUSSION THREAD

**Discussion by mdpletcher** [dtcenter/METplus#2365] [Installation]
I've been attempting to install and compile MET 11.0.0 onto my local directory within a CHPC environment, but I'm having issues with MET finding the netcdf library that I've configured and installed. 

MET is located at /home/user/software/pkg/miniconda3/met/MET-11.1.0

I've installed all of the required libraries at /home/user/software/pkg/miniconda3/met/libs and their respective paths are:

/home/user/software/pkg/miniconda3/met/libs/gsl-2.7.1
/home/user/software/pkg/miniconda3/met/libs/hdf5-1.10.10
/home/user/software/pkg/miniconda3/met/libs/NCEPLIBS-bufr-12.0.1

However, since the installation instructions state that the netcdf and netcdf-cxx files must be in the same folder, I created a file named 'netcdf' within the /libs directory that contains both libraries. Here are their respective paths:

/home/user/software/pkg/miniconda3/met/libs/netcdf/netcdf-cxx4-4.3.1
/home/user/software/pkg/miniconda3/met/libs/netcdf/netcdf-c-4.9.2

I'm operating in a .csh environment, so I create environment variables using setenv. I set the netcdf environment variable in the MET-11.1.0 directory (absolute path shown above) using the following command:

setenv MET_NETCDF /home/user/software/pkg/miniconda3/met/libs/netcdf/

Am I setting the environment variable incorrectly? 

Thanks,
Michael

**Comment by jprestop** [comment]
Hi Michael (@mdpletcher).  Thank you for all of the details in your post.

> Here are their respective paths:
/home/user/software/pkg/miniconda3/met/libs/netcdf/netcdf-cxx4-4.3.1
/home/user/software/pkg/miniconda3/met/libs/netcdf/netcdf-c-4.9.2
setenv MET_NETCDF /home/user/software/pkg/miniconda3/met/libs/netcdf/
Am I setting the environment variable incorrectly?

Thank you for recognizing that the netcdf and netcdf-cxx files must be in the same folder.  To explain further, the library files for netcdf-c and netcdf-cxx need to be in the same directory and the include files for netcdf-c and netcdf-cxx need to be in the same directory. So, for example, when installing both libraries you'd want to set the installation location with the --prefix option to the same location (e.g. `/home/user/software/pkg/miniconda3/met/libs/netcdf`), which would then put library files for both libraries in `/home/user/software/pkg/miniconda3/met/libs/netcdf/lib` and the include files in `/home/user/software/pkg/miniconda3/met/libs/netcdf/include`.  Then you could set `MET_NETCDF` to `/home/user/software/pkg/miniconda3/met/libs/netcdf` and MET would look for the corresponding `lib` and `include` directories.  

I hope that helps.  We do have a script, which we recommend using, that will compile MET and all of its external library dependencies for you, although I understand that you've already proceeded with installing many of these libraries.  You can find information about it on the [MET Dowloads Page](https://dtcenter.org/community-code/model-evaluation-tools-met/download) under "Sample Script For Compiling External Libraries And MET".  We are currently working on new installation instructions which will hopefully make the installation process more clear and smooth and plan to have those released in our next official release (MET-12.0.0).

Also, you mentioned that you are installing MET-11.0.0.  I wanted to be sure to point out that our latest official release is MET-11.1.0, which we would recommend installing over MET-11.0.0, as it came with lots of bugfixes.  You can take a look at the [release notes](https://met.readthedocs.io/en/main_v11.1/Users_Guide/release-notes.html) for MET-11.1.0 for a closer look at those bugfixes and enhancements, if you'd like, and you can obtain the latest release [here](https://github.com/dtcenter/MET/releases/tag/v11.1.0).

Please let us know if you have any other questions.  Please also let us know how the installation goes.

**Reply by mdpletcher** [reply]
Hi Julie,

Thanks so much for the detailed reply. I did not realize a compile script is provided on the downloads page, so that is very helpful.

I tried running the provided script to compile MET but received the following error at the end of the installation:

ERROR: Command returned with non-zero (2) status: make.test  >met.make_test.log2>&1

**Reply by jprestop** [reply]
Could you please pass along your met.make_test.log file as an attachment here?  You may need to rename the file to .txt.

**Comment by mdpletcher** [comment]
After looking through all files, I don't see the met.make_test.log file. Here's how I organized my directories for compilation:

Main MET folder: "/home/user/software/pkg/miniconda3/met"

Within that folder, I have the compile_MET_all.sh and install_met_env.generic files and I've attached the install_met_env.generic.

The tar files are located in "/home/user/software/pkg/miniconda3/met/tar_files" and the MET-11.1.0.tar.gz file is located in tar_files.

Michael

[install_met_env_generic.txt](https://github.com/dtcenter/METplus/files/12718417/install_met_env_generic.txt)

**Reply by jprestop** [reply]
@mdpletcher Thank you for this information.

I have updated the install_met_env.generic file in our repository.  Could you please grab this one instead?

> wget https://raw.githubusercontent.com/dtcenter/MET/main_v11.1/internal/scripts/installation/config/install_met_env.generic

This is an example environment configuration file to start from.  You'll need to update the file with the compiler and version of the compiler that you are using.  If you want the script to install all of the libraries, you can leave this section of the script alone:
```
# If users have already installed these libraries and would like to make use of
# them, uncomment out the export statements. If those pre-existing libraries are
# in the external_libs directory, no further edits are needed; however, users
# that have the pre-existing libraries not in the external_libs directory will
# need to update the paths to the appropriate location.
#export EXTERNAL_LIBS=${TEST_BASE}/external_libs
#export MET_GSL=${EXTERNAL_LIBS}
#export MET_BUFRLIB=${EXTERNAL_LIBS}/lib
#export BUFRLIB_NAME=-lbufr
#export LIB_JASPER=${EXTERNAL_LIBS}/lib
#export LIB_LIBPNG=${EXTERNAL_LIBS}/lib
#export LIB_Z=${EXTERNAL_LIBS}/lib
#export MET_GRIB2CLIB=${EXTERNAL_LIBS}/lib
#export MET_GRIB2CINC=${EXTERNAL_LIBS}/include
#export GRIB2CLIB_NAME=-lgrib2c
#export MET_HDF5=${EXTERNAL_LIBS}
#export MET_NETCDF=${EXTERNAL_LIBS}
```
If you want to use the libraries you have already installed, you'll need to uncomment the export statements and update the paths to the appropriate location.  Please let me know if you have any questions.

**Reply by mdpletcher** [reply]
Thanks for updating the install_met_env.generic file. I reran the configuration and I got the same error, but it created a met.make_test.log file this time. I've attached it for your reference.

[met.make.log](https://github.com/dtcenter/METplus/files/12719656/met.make.log)

**Reply by jprestop** [reply]
HI @mdpletcher.  Hrm.  That is some strange output there... Could you please attach the met.config.log file along with the updated generic configuration file that you are passing to the script?

**Reply by mdpletcher** [reply]
Yes, here they are.

[met.configure.log](https://github.com/dtcenter/METplus/files/12731593/met.configure.log)
[install_met_env.txt](https://github.com/dtcenter/METplus/files/12731602/install_met_env.txt)

**Reply by jprestop** [reply]
Thank you.  My apologies @mdpletcher - I asked for met.config.log, but I forgot that config.log is not prefixed with "met."  met.configure.log does not provide me with useful information. Could you please pass along config.log in the MET-12.0.0 directory?  Thanks!

**Reply by mdpletcher** [reply]
Ah, sorry about that. Here's that file.
[config.log](https://github.com/dtcenter/METplus/files/12731812/config.log)

**Reply by jprestop** [reply]
Thank you!  That is much more useful, and I do see a problem.

First, in your install_met_env.txt file I see that you have set:
```
export COMPILER=gnu_8.3.0
```
But, the config.log file shows version 8.5.0.  So please change that to:
```
export COMPILER=gnu_8.5.0
```
That is not the problem, however, just a side note.

The problem lies with the setting of MET_PYTHON_LD in the install_met_env.txt file:
```
export MET_PYTHON_LD=`python3-config --ldflags --embed`
```
MET-11.1.0 was developed using Python version 3.10.4. The `--embed` option referenced above only works with Python 3.8 and above, and I see you are using Python version 3.7.   That being said, you could probably use version 3.7 and it would likely work fine, but we do recommend using version 3.10.4.

If you do choose to proceed with version 3.7, you should change the values for MET_PYTHON_CC and MET_PYTHON_LD to the following:
```
export MET_PYTHON_CC="-I/uufs/chpc.utah.edu/common/home/u1325975/software/pkg/miniconda3/include/python3.7m"
export MET_PYTHON_LD=`python3-config --ldflags`
```
That should do the trick, however, sometimes the command `python3-config --ldflags` does not include the `-lpython3.m` in its output, which is needed for linking.  You can check to see if it is included in the new config.log file produced after rerunning.  For example, run:
```
grep MET_PYTHON_LD config.log
```
and check the value.  If it's not there, you'll need to reset MET_PYTHON_LD.  Please let me know if you have any other questions or problems.

If you don't want to reinstall all of the dependencies AND you've used the compilation script to install them, uncomment these lines in  your install_met_env.txt file:
```
#export EXTERNAL_LIBS=${TEST_BASE}/external_libs
#export MET_GSL=${EXTERNAL_LIBS}
#export MET_BUFRLIB=${EXTERNAL_LIBS}/lib
#export BUFRLIB_NAME=-lbufr
#export LIB_JASPER=${EXTERNAL_LIBS}/lib
#export LIB_LIBPNG=${EXTERNAL_LIBS}/lib
#export LIB_Z=${EXTERNAL_LIBS}/lib
#export MET_GRIB2CLIB=${EXTERNAL_LIBS}/lib
#export MET_GRIB2CINC=${EXTERNAL_LIBS}/include
#export GRIB2CLIB_NAME=-lgrib2c
#export MET_HDF5=${EXTERNAL_LIBS}
#export MET_NETCDF=${EXTERNAL_LIBS}
```
Please let me know if you have any questions.

**Reply by mdpletcher** [reply]
It looks like I was able to begin the compilation (getting closer!) but I received another error. I've attached the met.make.log file here that gives the error at the end.

[met.make.log](https://github.com/dtcenter/METplus/files/12732112/met.make.log)

Also, I ran grep MET_PYTHON_LD config.log and it gave the following output

ac_cv_env_MET_PYTHON_LD_set=set
ac_cv_env_MET_PYTHON_LD_value='-L/uufs/chpc.utah.edu/common/home/u1325975/software/pkg/miniconda3/lib/python3.7/config-3.7m-x86_64-linux-gnu -L/uufs/chpc.utah.edu/common/home/u1325975/software/pkg/miniconda3/lib -lpython3.7m -lcrypt -lpthread -ldl  -lutil -lm '
MET_PYTHON_LD='-L/uufs/chpc.utah.edu/common/home/u1325975/software/pkg/miniconda3/lib/python3.7/config-3.7m-x86_64-linux-gnu -L/uufs/chpc.utah.edu/common/home/u1325975/software/pkg/miniconda3/lib -lpython3.7m -lcrypt -lpthread -ldl  -lutil -lm '

My edited install_met_env.txt file is also attached.

[install_met_env.txt](https://github.com/dtcenter/METplus/files/12732114/install_met_env.txt)

Thanks so much.

**Reply by jprestop** [reply]
Hi @mdpletcher.  Thank you for attaching these files.  I see the following error in your met.make.log file:

> lto1: fatal error: bytecode stream in file ‘/uufs/chpc.utah.edu/common/home/u1325975/software/pkg/miniconda3/lib/python3.7/config-3.7m-x86_64-linux-gnu/libp\
ython3.7m.a’ generated with LTO version 6.2 instead of the expected 7.3

which reminded me of a [Discussion post](https://github.com/dtcenter/METplus/discussions/2012) from earlier in the year where another user experienced a similar LTO error.  Other users have also encountered this problem before. In those cases, the miniconda Python gcc was a different version compared with what they were using to compile MET and its dependent libraries. To get a successful compilation, the users switched to using the system's installed default version of Python instead to get a successful installation. Is that an option for you? If not, are there gcc, g++, and gfortran executables within your miniconda area that you could point to (by setting the values for CC, CXX, FC, F77, and F90 in the install_met_env.txt) instead to try building with those?  If you do set those variables, you will need to recompile the external libraries, along with MET, and as the external libraries should be built with the same compiler that MET is built with.

**Reply by jprestop** [reply]
Hi @mdpletcher.  I just wanted to follow up on this Discussion and see if you were able to make any progress?

**Reply by mdpletcher** [reply]
Hi @jprestop 
Sorry I haven't gotten back to you. I'm currently putting this issue on hold for now but I plan to get back to it soon. Thank you so much for all of your help thus far.

**Reply by jprestop** [reply]
Thank you for the update @mdpletcher.  That's no problem at all.  I just wanted to check in.

**Reply by jprestop** [reply]
Hi @mdpletcher.  I just thought I'd follow up again and see if you have been able to make any progress?

**Reply by jprestop** [reply]
Hi @mdpletcher.  It's been a while since I checked in on this work.  I thought I'd check in and see if you were able to make any progress on this problem.

**Reply by mdpletcher** [reply]
Hi Julie, so sorry I haven't replied to you yet. I got caught up with a different research project that delayed my installation of MET. I'm about finished that project, however, and you should expect to hear an update on this issue in the next month or so. Thanks so much for continuing to follow up!

**Reply by jprestop** [reply]
That sounds great, @mdpletcher.  Thank you!
